### ISSUE TITLE
Sniff SARIF version before attempting to deserialize

### ISSUE THREAD

**Issue by lgolding** [Microsoft/sarif-sdk#1165]
The SARIF spec says (§3.12.2):

> Although the order in which properties appear in a JSON object value is not semantically significant, the version property **SHOULD** appear first.
> NOTE: This will make it easier for parsers to handle multiple versions of the SARIF format if new versions are defined in the future.

But the SARIF SDK today does not take the hint 😊 Currently, the recommended coding pattern is to use Newtonsoft.Json to blindly deserialize the SARIF file into an object model, and that can fail if the file's SARIF version is not compatible with the object model supported by the SDK.

To solve this problem, we could introduce a wrapper in the SDK, for example, a static method on the `SarifLog class` like
```
SarifLog SarifLog.ReadFrom(string path);
```

The wrapper would sniff the version property and throw a distinguished exception with a message like
```
This program can read SARIF files that conform to SARIF version {0}.
The SARIF log file '{1}' cannot be read because it conforms to SARIF version {2}.
```
... and we would update the documentation to recommend this wrapper.

**Comment by lgolding** [comment]
@shaunm-msft, @michaelcfanning FYI.

**Comment by michaelcfanning** [comment]
I agree this is an important issue. Generally, requiring people to locate and load files through a wrapper is not a recipe for consistent success. This is one of the reasons why we've worked so hard to eliminate the requirement to provide a contract resolver (which could have been handled in the way you describe above).

I'm not arguing against this suggestion. It'd be nice to encourage people to call through a wrapper API that would allow the SDK to do some magic. We could automatically invoke the transformation code here, for example.

We should add a new validator that checks the position of the version property as well.
